### PR TITLE
table, executor: let `SplitIndexValue` work with clustered index table.

### LINKED ISSUE #49995: executor: enhance global index for clustered index table
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```SQL
set tidb_enable_global_index=true;
create table pt (
    a int, b int, c int, d int default 0, 
    primary key (a, b) clustered, 
    unique key uidx(c)) partition by range(a) ( 
        PARTITION p0 VALUES LESS THAN (3), 
        PARTITION p1 VALUES LESS THAN (6), 
        PARTITION p2 VALUES LESS THAN (9), 
        PARTITION p3 VALUES LESS THAN (20) 
);
insert into pt(a,b,c) values(1,1,1), (2,2,2), (3,3,3), (4,4,4), (5,5,5), (6,6,6), (7,7,7), (8,8,8), (9,9,9), (10,10,10);
analyze table pt;
select c from pt where c = 1;
select * from pt where c in (1,2,3);
```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
for point get: 1,1,1;
for batch point get: 1,1,1;2,2,2;3,3,3;
### 3. What did you see instead (Required)
<img width="463" alt="image" src="https://github.com/pingcap/tidb/assets/20724552/1dc82131-7eb1-4918-8efd-321f1fabe835">
### 4. What is your TiDB version? (Required)
mysql> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                 |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.6.0-alpha-629-g89fc4403f3
Edition: Community
Git Commit Hash: 89fc4403f3a05d59a6793bf1965dc2281e3fb29f
Git Branch: local/fix_global_index_under_clustered_index
UTC Build Time: 2024-01-03 02:59:45
GoVersion: go1.21.0
Race Enabled: false
Check Table Before Drop: false
Store: unistore |
+--------------------------------------------------------
<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #49995

Problem Summary: Ref PR https://github.com/pingcap/tidb/pull/50013, remove `insert on duplicate` part. /cc @L-maple 

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/table/tables/index.go [hunk 1]
```diff
@@ -412,15 +412,12 @@ func (c *index) Delete(ctx table.MutateContext, txn kv.Transaction, indexedValue
 		// Only when pid in value equals to c.physicalID, the key can be deleted.
 		if c.idxInfo.Global {
```
Comment (L-maple) [inline pkg/table/tables/index.go:413]: L413~L425 are useless now, which can be deleted. ref: https://github.com/pingcap/tidb/pull/52442?w=1

File: pkg/tablecodec/tablecodec.go [hunk 2]
```diff
@@ -1730,8 +1719,23 @@ type IndexValueSegments struct {
...
+		}
+		// For IndexValueVersion0
+		return splitIndexValueForIndexValueVersion0(value)
+	}
+	// For IndexValueForClusteredIndexVersion1
+	return splitIndexValueForClusteredIndexVersion1(value)
+}
+
+// splitIndexValueForIndexValueVersion0 splits index value into segments.
+func splitIndexValueForIndexValueVersion0(value []byte) (segs IndexValueSegments) {
```
Comment (L-maple) [inline pkg/tablecodec/tablecodec.go:1738]: not get the name `splitIndexValueForIndexValueVersion0` logic behind? this is a extensible index value format.

Comment (Defined2014) [reply pkg/tablecodec/tablecodec.go:1738]: See https://github.com/pingcap/tidb/blob/2d552c00343ca9a457fcda2924c3940f7f46add0/pkg/tablecodec/tablecodec.go#L1408, we have three different index value types.

**Comment by L-maple** [comment]
`SplitIndexValue` is really nice as a facade, thumb up for you @Defined2014

File: pkg/tablecodec/tablecodec.go (hunk 2 above)
Comment (L-maple) [reply pkg/tablecodec/tablecodec.go:1738]: okok

File: pkg/table/tables/index.go (hunk 1 above)
Comment (Defined2014) [reply pkg/table/tables/index.go]: Done, thx.

**Review by mjonss — COMMENTED** [review]
LGTM, with some questions:
- So the default of creating a partitioned table with a global index, will still be non-clustered?
- Should we also create a PR for supporting clustered tables for DDL, which seems to be disabled [here](https://github.com/pingcap/tidb/blob/f0af7766a2c2f9c779315b9fa75283c1f7a182ed/pkg/ddl/partition.go#L3895)?

**Comment by Defined2014** [comment]
> * So the default of creating a partitioned table with a global index, will still be non-clustered?
The problem is we run `set @@tidb_enable_clustered_index='int_only'` in mysql-tester, [link](https://github.com/pingcap/mysql-tester/blob/aa83826f27fddbb26252d0b1c780ad873a43afd5/src/main.go#L189)

> Should we also create a PR for supporting clustered tables for DDL, which seems to be disabled [here](https://github.com/pingcap/tidb/blob/f0af7766a2c2f9c779315b9fa75283c1f7a182ed/pkg/ddl/partition.go#L3895)?

This line means we couldn't create an index is global and also is a primary key(clustered or nonclustered). Because a primary key should contains partition column which global index not contain it. But if table has a clustered index, it's ok. I thin this line makes error message more cleaner.

**Review by zimulala — APPROVED** [review]
LGTM

**Review by mjonss — APPROVED** [review]

**Pull request merged**